Core/Scripts Boss Lady DeathWhisper improvements

sirikfoll opened the pull request:
**Changes proposed**:
- Corrected the Adds behavior
- Timers fixed
- EventMap->Taskscheduler
- DoCast(me, ->DoCastSelf
- Mind Controlled player size increase
- Aggro reset on transition to phase two
- Vengefull Shade behavior, use proper summon spell
- Changed spell spell_gen_creature_permanent_feign_death to remove the flags at aura removal
- General clean-up, optimizations

**Target branch(es)**: 335/6x

**Issues addressed**: Fixes #5549 #7737

**Tests performed**: Built and tested(alone testing)

**Known issues and TODO list**:
- [x] Correct spell used to handle Dominate Mind players scale
- [x] Vengefull Shade speed and time until start attack fixed

P-Kito commented:
Urgh, I really need to code something with TaskScheduler. Feels so odd.

Nice job :+1:

P-Kito commented on src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp:
```diff
@@ -815,12 +791,15 @@ class npc_vengeful_shade : public CreatureScript
 
             void Reset() override
             {
                 me->AddAura(SPELL_VENGEFUL_BLAST_PASSIVE, me);
```
Vengeful Shades should stand still for a second, then follow its target. Please also check the speed with videos - I remember it being wrong.

sirikfoll replied:
A lot of summoned creatures should stand still for a little time, I think this could be some generic thing

sirikfoll replied:
@P-Kito  The speed was indeed wrong, and he waits 2 seconds from summon to start attack, fixed ^^

P-Kito commented on src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp:
```diff
@@ -373,148 +395,96 @@ class boss_lady_deathwhisper : public CreatureScript
+                        .Schedule(Seconds(6), Seconds(9), GROUP_TWO, [this](TaskContext touch)
+                        {
+                            if (me->GetVictim())
+                                me->AddAura(SPELL_TOUCH_OF_INSIGNIFICANCE, me->EnsureVictim());
```
AddAura?

sirikfoll replied:
yes, there's no Cast in sniffs

sirikfoll commented:
I have found this crash using this script, dunno reproduce steps and can't provide a full debug crashLog
```
binind = 229
#0  0x0000000001af9d6a in arena_salloc (demote=false, ptr=0x1cfb8f8 <vtable for SpellScript+16>) at /wow/br15/dep/jemalloc/include/jemalloc/internal/arena.h:1018
#1  isalloc (demote=false, ptr=0x1cfb8f8 <vtable for SpellScript+16>) at /wow/br15/dep/jemalloc/include/jemalloc/internal/jemalloc_internal.h:880
#2  ifree (ptr=0x1cfb8f8 <vtable for SpellScript+16>) at /wow/br15/dep/jemalloc/src/jemalloc.c:1228
#3  free (ptr=0x1cfb8f8 <vtable for SpellScript+16>) at /wow/br15/dep/jemalloc/src/jemalloc.c:1308
#4  0x000000000102da68 in __gnu_cxx::new_allocator<std::_List_node<ObjectGuid> >::deallocate (this=<optimized out>, __p=<optimized out>) at /usr/include/c++/5/ext/new_allocator.h:110
#5  std::__cxx11::_List_base<ObjectGuid, std::allocator<ObjectGuid> >::_M_put_node (__p=<optimized out>, this=<optimized out>) at /usr/include/c++/5/bits/stl_list.h:396
#6  std::__cxx11::_List_base<ObjectGuid, std::allocator<ObjectGuid> >::_M_clear (this=<optimized out>) at /usr/include/c++/5/bits/list.tcc:79
#7  std::__cxx11::list<ObjectGuid, std::allocator<ObjectGuid> >::clear (this=<optimized out>) at /usr/include/c++/5/bits/stl_list.h:1368
#8  boss_lady_deathwhisper::boss_lady_deathwhisperAI::Initialize (this=<optimized out>) at /wow/br15/src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp:204
#9  boss_lady_deathwhisper::boss_lady_deathwhisperAI::Reset (this=0x7fff3cfbb280) at /wow/br15/src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp:216
#10 0x000000000173d515 in Creature::Respawn (this=this@entry=0x7ffe616b9000, force=force@entry=false) at /wow/br15/src/server/game/Entities/Creature/Creature.cpp:1742
#11 0x000000000173dfca in Creature::Update (this=0x7ffe616b9000, diff=92) at /wow/br15/src/server/game/Entities/Creature/Creature.cpp:524

Aokromes commented:
@shauren any idea on https://github.com/TrinityCore/TrinityCore/pull/17389#issuecomment-229502905

Killyana commented:
For the dominate mind scale+red color: I wasn't able to find any thing on sniffs, but by looking on dbc spells you can see: 71282 and 71290, probably one is used for the scale and the other for the color.

sirikfoll commented on src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp:
```diff
@@ -653,64 +590,90 @@ class npc_cult_fanatic : public CreatureScript
+                    {
+                        DoCastVictim(SPELL_NECROTIC_STRIKE);
+                        necrotic_strike.Repeat(Seconds(17));
+                    }); 
```
just a reminder, whitespace.

sirikfoll commented on src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp:
```diff
@@ -1004,50 +1006,93 @@ class spell_deathwhisper_mana_barrier : public SpellScriptLoader
-                    if (Creature* ladyDeathwhisper = ObjectAccessor::GetCreature(*player, instance->GetGuidData(DATA_LADY_DEATHWHISPER)))
-                        ladyDeathwhisper->AI()->DoAction(ACTION_START_INTRO);
+            PrepareSpellScript(spell_deathwhisper_summon_spirits_SpellScript);
+            
```
just a reminder, whitespace.

sirikfoll commented:
@Killyana I used only the spell 71290 to handle the scale, the color is already added by the main spell, thanks :)
Waiting to hear if anyone knows something about the crashes.
As for the boss functionality, i think it's everything finished.
Feedback will be appreciated ^^

sucklifex commented:
http://paste.ubuntu.com/23052536/

sucklifex commented:
after this update sometimes Lady gets stuck and will do noting

Keader commented:
@sucklifex have you steps for reproduce it?

sucklifex commented:
@Keader not yet, it happens very rarely

Treeston commented:
@sirikfoll In `JustSummoned` for `NPC_VENGEFUL_SHADE` you never check whether `_nextVengefulShadeTargetGUID` is empty. Both `::front()` and `::pop_front` cause UB when called on an empty list.

sirikfoll commented:
Last commit can probably fix https://github.com/TrinityCore/TrinityCore/pull/17389#issuecomment-229502905

Aokromes commented:
@sirikfoll 

This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.

tkrokli commented:
Specifically, merge the changes Treeston made in the /src/server/scripts/ folders & files. Just take care with your own files so you don't lose any content. :-)

sirikfoll committed:
Core/Script Boss Lady DeathWhisper improvements

Corrected the Adds behavior
Timers fixed
EventMap->Taskscheduler
DoCast(me, ->DoCastSelf
Use spell 71290 from spell_dbc to handle Dominate Mind scale
Aggro reset on transition to phase two
Vengefull Shade behavior, use proper summon spell
Changed spell spell_gen_creature_permanent_feign_death to remove the flags at aura removal
General clean-up, optimizations

Demonid commented:
I believe this should be merged, if crashes are still a thing it would be easier for tc users to post crashes and fix the remaining issues if there are still some.

Treeston commented on src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp:
```diff
@@ -943,25 +938,25 @@ class npc_darnavan : public CreatureScript
                         case EVENT_DARNAVAN_MORTAL_STRIKE:
                             DoCastVictim(SPELL_MORTAL_STRIKE);
-                            _events.ScheduleEvent(EVENT_DARNAVAN_MORTAL_STRIKE, urand(15000, 30000));
+                            _events.ScheduleEvent(EVENT_DARNAVAN_MORTAL_STRIKE, randtime(Seconds(15), Seconds(30)));
```
EventMap overload means you no longer need `randtime` for scheduling. Also, this should be `.Repeat`.

Treeston commented on src/server/scripts/Northrend/IcecrownCitadel/boss_lady_deathwhisper.cpp:
```diff
@@ -736,76 +695,87 @@ class npc_cult_adherent : public CreatureScript
+                                me->RemoveAurasDueToSpell(SPELL_PERMANENT_FEIGN_DEATH);
+                                me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_STUNNED | UNIT_FLAG_UNK_29 | UNIT_FLAG_NOT_SELECTABLE);
+                                me->SetReactState(REACT_AGGRESSIVE);
+                                DoCastSelf(SPELL_SHORUD_OF_THE_OCCULT);
```
`SHORUD` -> `SHROUD`

sirikfoll committed:
Minor Improvements

Keader commented:
Thanks @SnapperRy  ^^

tkrokli commented:
... and thanks to sirikfoll for creating the PR content. 👍
